sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/commands.scala

```diff
@@ -244,34 +244,65 @@ private[sql] case class InsertIntoHadoopFsRelation(
     }
 
     def writeRows(taskContext: TaskContext, iterator: Iterator[InternalRow]): Unit = {
+      // Track which rows have been output to disk so that if a data sort is necessary mid-write,
+      // we don't end up outputting the same data twice
+      val writtenRows: HashSet[InternalRow] = new HashSet[InternalRow]
+
+      // Flag to track whether data has been sorted in which case it's safe to close previously
+      // used outputWriters
+      var sorted: Boolean = false
+
       // If anything below fails, we should abort the task.
       try {
         writerContainer.executorSideSetup(taskContext)
 
-        // Projects all partition columns and casts them to strings to build partition directories.
-        val partitionCasts = partitionOutput.map(Cast(_, StringType))
-        val partitionProj = newProjection(codegenEnabled, partitionCasts, output)
-        val dataProj = newProjection(codegenEnabled, dataOutput, output)
+        // Sort the data by partition so that it's possible to use a single outputWriter at a
+        // time to process the incoming data
+        def sortRows(iterator: Iterator[InternalRow]): Iterator[InternalRow] = {
+          // Sort by the same key used to look up the outputWriter to allow us to recyle the writer
+          iterator.toArray.sortBy(writerContainer.computePartitionPath).toIterator
```

Comment:
I think the items will be loaded into memory lazily by iterator, that's the advantage of Iterator than Array.